Update the invite page to include a link to the Slack workspace

accorvin opened the pull request:
This addresses #35 

I've additionally updated the development and contributing guides based on some issues I found.

accorvin committed:
Update Ruby version in dev instructions

The Gemfile requires Ruby version 2.7.7. Updating the development
prerequisites accordingly.

accorvin committed:
Add a link to Slack workspace on invitation page

accorvin committed:
Update contribution guide to match changelog format

accorvin commented:
@dblock in its current state this breaks 3 of the spec tests. I'm hoping you can help me get them working. The tests don't find the field named `name` to fill in the data for, which is a little baffling to me.

dblock commented on public/invite.html.erb:
```diff
@@ -13,6 +13,7 @@
   </p>
   <p>
     <h1 id='team_caption'>Get Invited to Team <u><span id='team_name'></span></u></h1>
+    <span>If you have previously joined this team, click <a id='team_link'>here</a> to be redirected</span>
```
To be super nitpicky, let's call this `team_workspace_url` to be consistent with the other field names.

Add a period since it's a full sentence.

dblock commented on public/invite.html.erb:
```diff
@@ -32,6 +34,9 @@
 
         $('#team_name').text(team.name);
+
+        a = document.getElementById('team_link');
```
Since we use JQuery above, let's do `$('#team_link').attr("href", newUrl)`

dblock commented on public/js/common.js:
```diff
@@ -36,6 +36,9 @@ $(document).ready(function() {
         }
       }
     }
+    if (message == "already_in_team") {
```
Move this into the `else if` above just like for `code_already_used`?

dblock commented on slack-invite/models/team.rb:
```diff
@@ -62,6 +64,11 @@ def inform!(message)
 
+  def workspace_url
+    team_info = slack_client.team_info
+    team_info.to_hash['team']['url']
```
I would break this into two methods:

```
def team_info
    @team_info ||= slack_client.team_info
end

def workspace_url
   team_info&.team.url
end
```

(See if `to_hash...` is unnecessary and whether the above works).

... but see below - we don't want to have to call the API every time.

dblock reviewed (changes requested):
Cool! I have a bunch of rather nitpicky comments. Hang in there with me. 

We need to write a (integration) test for this too. LMK if you need help.

dblock commented:
> @dblock in its current state this breaks 3 of the spec tests. I'm hoping you can help me get them working. The tests don't find the field named `name` to fill in the data for, which is a little baffling to me.

I did a few things to debug this.

1. I made sure my tests pass on master `git checkout master ; rake`.
2. I made sure tests don't pass on this branch.
3. Reproduced with one test, `rspec ./spec/integration/invite_spec.rb:13`.
4. Noticed that the page loaded `http://127.0.0.1:9293/invite?team_id=T0` is blank. The server fails with a 500.
5. Guessed that the reason is that `<%= team && team.workspace_url %>` is failing. This is because there's no slack key, the specs are now calling `workspace_url` which is making an API call to slack.

This is one way to stub it.

```ruby
  before do
    allow_any_instance_of(Slack::Web::Client).to receive(:team_info).and_return(
      team: {
        url: 'https://example.slack.com'
      }
    )
  end
```

Now, I think we should not be calling the Slack API every time this page is loaded, this is an easy way to get blocked by Slack (one can DOS this page by guessing the URL), this information doesn't change often, if at all, and we don't want to do an external call for latency either. Add a field to `Team` called `workspace_url` and we can populate its value on create, so something like this.

```ruby
field :workspace_url, type: String

def team_info
    @team_info ||= slack_client.team_info
end

before_save :update_team_info!

def update_team_info!
   self.workspace_url ||= team_info&.team&.url   
end
```

Existing teams can be backfilled once manually.

For tests try using `, vcr: { cassette_name: 'slack/team_info' }` where appropriate instead of stubbing like I suggested above. I think we used to get team info before, the team_info.yml is in the source code already that has a response for that API call.

Tests that `Fabricate(:team)` can have a value for `workspace_url` added into the fabricator, `team_fabricator.rb`.

accorvin committed:
Address review comments and fix tests

accorvin committed:
Add integration tests for workspace_url

accorvin commented:
@dblock I've finally had time to get back to this. I've addressed your comments and gotten the existing tests working.

I've managed to add an integration test confirming that the workspace_url is linked on the page. I'm trying to add another integration test for the message that is displayed when a user is already in the workspace, but I'm having trouble getting that one working. I'm hoping you can give me a pointer there.

Right now I've split this up into a few commits just so you can see my recent changes more easily. I plan to squash my commits prior to this being merged.

dblock commented:
> I've managed to add an integration test confirming that the workspace_url is linked on the page. I'm trying to add another integration test for the message that is displayed when a user is already in the workspace, but I'm having trouble getting that one working. I'm hoping you can give me a pointer there.

I think you've confused some workflows. The user goes to the invite page and types their name and email and clicks submit. Then an invitation is created and either `users_admin_invite` is called, or, if the invitation requires approval an admin needs to click something and then `users_admin_invite` is called upon approval. So is the error `already_in_team` returned from the Slack `users_admin_invite` call? Test that first. If that's the case, in the test you've added you expect `Invitation#approve!` to be called, which only happens on approval, which is not the context of that test. Expect `users_admin_invite` to be called and raise `already_in_team`.

The other thing to notice is that it's 2 scenarios: one where the user self-invites (user gets the error message), and the other where the admin approves and then the admin gets a message. I worry that if we do this it becomes a way to check whether any user is a member of a workspace, which I think is not a good idea. Imagine a Slack group called "criminal gang", I can see if dblock[at]dblock[dot]org is a member by filling this form 🤔 

Either way, I think you should reduce this PR to just the adding a link part, your original goal, and let's get that merged? Then you can take your time trying to make the above work.

dblock commented on public/js/common.js:
```diff
@@ -33,6 +33,8 @@ $(document).ready(function() {
           message = 'The code returned from the OAuth workflow was invalid.'
         } else if (message == 'code_already_used') {
           message = 'The code returned from the OAuth workflow has already been used.'
+        } else if (message == "already_in_team") {
```
Remove this from this PR.

dblock commented on spec/integration/invite_spec.rb:
```diff
@@ -10,6 +10,10 @@
       let!(:team) { Fabricate(:team, admin_token: 'token') }
+      it "includes a link to the team's slack workspace" do
+        visit "/invite?team_id=#{team.team_id}"
+        expect(page).to have_link(nil, href: team.workspace_url)
```
Can we be more specific that the page has something like `If you have previously joined this team, click <a id='team_workspace_url'>here</a> to be redirected.`? Maybe check that the `href` on `team_workspace_url` matches this URL.

accorvin commented:
My needs around this have changed and I won't be able to get back to this PR any time soon, so closing it.

dblock commented:
A version of this was added in https://github.com/vestris/slack-invite/commit/fff5ea00f122025d81c7eb6da0b848abb7cb5bf2.
